File: 5.最长回文子串.c
/*
 * @lc app=leetcode.cn id=5 lang=c
 *
 * [5] 最长回文子串
 */

// @lc code=start

int getMax(char *s, int mid)
{
  int i, j, flag = 0;
  if (strlen(s) % 2 == 0)
  {
    i = mid;
    j = i + 1;
  }
  else
    i = j = mid;

  while (i >= 0 && j <= strlen(s) && s[i] == s[j])
  {
    i--;
    j++;
    flag++;
  }

  if (flag)
    return j - i - 1;
  else
    return j - i;
}

char *longestPalindrome(char *s)
{
  // 中心检索
  int i, len, max = 0, mid, j, left, right;
  for (i = 0; i < strlen(s); i++)
  {
    len = getMax(s, i);
    if (max < len)
    {
      max = len;
      mid = i;
    }
  }

  if (strlen(s) % 2)
  {
    left = mid - max / 2;
    right = mid + max / 2;
  }
  else
  {
    left = mid;
    right = mid + max / 2;
  }

  printf("%d,%d\n", mid, max);
  // printf("%d,%d", mid - max / 2, mid + max / 2);
  char *res;
  res = (char *)malloc(sizeof(char) * (max + 1));
  for (i = left, j = 0; i <= right; i++)
    res[j++] = s[i];

  res[j] = '\0';
  return res;
}
// @lc code=end

File: store/5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  // let dp = [],
  //   i,
  //   j,
  //   max,
  //   left,
  //   right;
  // max = 0;
  // dp = Array(s.length)
  //   .fill()
  //   .map(() => Array(s.length));
  // for (i = s.length - 1; i >= 0; i--) {
  //   for (j = 0; j < s.length; j++) {
  //     dp[i][j] = false;
  //   }
  // }
  // for (i = s.length - 1; i >= 0; i--) {
  //   for (j = 0; j < s.length; j++) {
  //     if (s[i] === s[j]) {
  //       if (j - i <= 1) dp[i][j] = true;
  //       else if (dp[i + 1][j - 1] === true) dp[i][j] = true;
  //     }
  //     if (dp[i][j] && j - i + 1 > max) {
  //       max = j - i + 1;
  //       left = i;
  //       right = j;
  //     }
  //   }
  // }
  // return s.slice(left, right + 1);

  let dp = new Array(s.length)
      .fill()
      .map(() => new Array(s.length).fill(false)),
    max = 0,
    left,
    right;

  for (let i = s.length - 1; i > -1; i--) {
    for (let j = i; j < s.length; j++) {
      if (s[i] === s[j]) {
        if (j - i <= 1) dp[i][j] = true;
        else if (dp[i + 1][j - 1]) dp[i][j] = true;
      }
      if (dp[i][j] && j - i + 1 > max) {
        max = j - i + 1;
        left = i;
        right = j;
      }
    }
  }

  return s.slice(left, right + 1);
};
// @lc code=end
